vm/actor/src/builtin/miner/bitfield_queue.rs

```diff
@@ -0,0 +1,147 @@
+        &mut self,
+        epoch: ChainEpoch,
+        values: &[u64],
+    ) -> Result<(), Box<dyn StdError>> {
+        self.add_to_queue(epoch, &values.iter().map(|&i| i as usize).collect())
+    }
+
+    /// Cut cuts the elements from the bits in the given bitfield out of the queue,
+    /// shifting other bits down and removing any newly empty entries.
+    ///
+    /// See the docs on `BitField::cut` to better understand what it does.
+    pub fn cut(&mut self, to_cut: &BitField) -> Result<(), String> {
+        let mut epochs_to_remove = Vec::<u64>::new();
+
+        self.amt
+            .for_each_mut(|epoch, bitfield| {
+                *bitfield = bitfield.cut(to_cut);
+
+                if bitfield.is_empty() {
+                    epochs_to_remove.push(epoch);
+                }
+
+                Ok(())
+            })
+            .map_err(|e| format!("failed to cut from bitfield queue: {:?}", e))?;
+
+        // TODO: batch delete
+        for epoch in epochs_to_remove {
+            self.amt.delete(epoch).map_err(|e| {
+                format!("failed to remove empty epochs from bitfield queue: {:?}", e)
+            })?;
+        }
+
+        Ok(())
+    }
+
+    pub fn add_many_to_queue_values(
+        &mut self,
+        values: &HashMap<ChainEpoch, Vec<u64>>,
+    ) -> Result<(), Box<dyn StdError>> {
+        // Update each epoch in-order to be deterministic.
+        // Pre-quantize to reduce the number of updates.
+
+        let mut quantized_values = HashMap::<ChainEpoch, Vec<u64>>::new();
+        let mut updated_epochs = Vec::<ChainEpoch>::new();
+
+        for (&raw_epoch, entries) in values {
+            let epoch = self.quant.quantize_up(raw_epoch);
+            updated_epochs.push(epoch);
+            quantized_values.entry(epoch).or_default().extend(entries);
+        }
+
+        updated_epochs.sort_unstable();
+
+        for epoch in updated_epochs {
+            self.add_to_queue_values(epoch, &quantized_values.remove(&epoch).unwrap_or_default())?;
+        }
+
+        Ok(())
+    }
+
+    /// Removes and returns all values with keys less than or equal to until.
+    /// Modified return value indicates whether this structure has been changed by the call.
+    pub fn pop_until(&mut self, until: ChainEpoch) -> Result<(BitField, bool), Box<dyn StdError>> {
+        let mut popped_values = Vec::<BitField>::new();
+        let mut popped_keys = Vec::<u64>::new();
+
+        self.amt.for_each_while(|epoch, bitfield| {
+            if epoch as ChainEpoch > until {
+                // break
+                return Ok(false);
+            }
+
+            popped_keys.push(epoch as u64);
+            popped_values.push(bitfield.clone());
+            Ok(true)
+        })?;
+
+        if popped_keys.is_empty() {
+            // Nothing expired.
+            return Ok((BitField::new(), false));
+        }
+
+        // TODO: batch delete
+        for key in popped_keys {
+            self.amt.delete(key)?;
+        }
```

Comment:
maybe worth just adding the bulk delete function on the amt and calling it with this logic. Just to future proof it so if it does get optimized later this doesn't get missed

Reply:
Yep, great call